critiquebrainz/db/oauth_grant.py

```diff
@@ -0,0 +1,141 @@
+              RETURNING id
+        """), grant)
+        grant["id"] = result.fetchone()[0]
+    return grant
+
+
+def list_grants(*, client_id=None, code=None, limit=1, offset=None):
+    """Returns the list of OAuth Grants.
+
+    Args:
+        client_id (str, optional): ID of the OAuth Client.
+        code (str, optional): The authorization code returned from authorization request.
+        limit (int, optional): Max number of grants to be returned by this method (default=1).
+        offset (int): Offset that can be used in conjunction with the limit.
+
+    Returns:
+        List of dictionaries with the following structure:
+        {
+            "id": (int),
+            "client_id": (str),
+            "code": (str),
+            "expires": (datetime),
+            "redirect_uri": (str),
+            "user_id": (uuid),
+            "scopes": (str),
+        }
+    """
+    filters = []
+    filter_data = {}
+
+    if client_id is not None:
+        filters.append("client_id = :client_id")
+        filter_data["client_id"] = client_id
+    if code is not None:
+        filters.append("code = :code")
+        filter_data["code"] = code
+    filterstr = str()
+    if filters:
+        filterstr = " AND ".join(filters)
+        filterstr = " WHERE " + filterstr
+
+    filter_data["limit"] = limit
+    filter_data["offset"] = offset
+
+    with db.engine.connect() as connection:
+        results = connection.execute(sqlalchemy.text("""
+            SELECT id,
+                   client_id,
+                   code,
+                   expires,
+                   user_id,
+                   redirect_uri,
+                   scopes
+              FROM oauth_grant
+              {where_clause}
+             LIMIT :limit
+            OFFSET :offset
+        """.format(where_clause=filterstr)), filter_data)
+        rows = results.fetchall()
+    return [dict(row) for row in rows]
+
+
+def delete(*, client_id, code):
+    """Delete an OAuth Token.
+
+    Args:
+        client_id (str): ID of the OAuth Client.
+        code (str): The authorization code returned from authorization request.
+    """
+    with db.engine.connect() as connection:
+        connection.execute(sqlalchemy.text("""
+            DELETE
+              FROM oauth_grant
+             WHERE client_id = :client_id
+               AND code = :code
+        """), {
+            "client_id": client_id,
+            "code": code,
+        })
+
+
+def get_scopes(grant_id):
+    """Returns the scopes of an application.
+
+    Args:
+        grant_id (int): ID of the OAuth Grant.
+    Returns:
+        scopes: (list).
+    """
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            SELECT scopes
+              FROM oauth_grant
+             WHERE id = :grant_id
+        """), {
+            "grant_id": grant_id,
+        })
+        scopes = result.fetchone()
+        if not scopes:
+            return list()
```

Comment:
There's the same issue of uncertainty about whether an item exists in the database as I pointed out in the other PR.